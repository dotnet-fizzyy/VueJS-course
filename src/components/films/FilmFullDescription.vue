<template>
    <div :class="$style.root">
        <!--v-lazy is registered as plugin-->
        <img v-lazy="posterUrl" :class="$style.poster" :alt="posterAlt" />

        <div :class="$style['film-description-container']">
            <div :class="$style['name-container']">
                <span :class="$style.name">{{ name }}</span>
                <span :class="$style.rating">{{ rating | precision }}</span>
            </div>

            <span :class="$style['short-description']">{{ shortDescription }}</span>

            <div :class="$style['summary-container']">
                <film-description-item :key-label="releaseYear | string" additional-label="year" />
                <film-description-item :key-label="runtime | round('up')" additional-label="min" />
            </div>

            <span :class="$style.description">
                {{ fullDescription }}
            </span>
        </div>
    </div>
</template>

<script lang="ts">
import FilmDescriptionItem from '@/components/films/FilmDescriptionItem.vue';
import Vue from 'vue';
import { getAppSettings } from '@/utils/appSettings';

export interface FilmFullDescriptionProps {
    name: string;
    releaseYear: number;
    runtime: number;
    rating: number;
    shortDescription: string;
    fullDescription: string;
    posterUrl?: string;
    posterAlt?: string;
}

export default Vue.extend({
    name: 'FilmFullDescription',
    components: { FilmDescriptionItem },
    props: {
        name: {
            type: String,
            required: true,
        },
        posterUrl: {
            type: String,
            default: getAppSettings().defaultPosterUrl,
        },
        posterAlt: {
            type: String,
            default: getAppSettings().defaultPosterAlt,
        },
        shortDescription: {
            type: String,
            required: true,
        },
        fullDescription: {
            type: String,
            required: true,
        },
        releaseYear: {
            type: Number,
            required: true,
        },
        rating: {
            type: Number,
            required: true,
        },
        runtime: {
            type: Number,
            required: true,
        },
    },
});
</script>

<style lang="scss" module scoped>
@import '../../styles/fonts';

.root {
    width: 100%;
    display: flex;
    flex-direction: row;
}

.poster {
    width: 280px;
    height: 390px;
}

.film-description-container {
    width: 100%;
    margin-left: 60px;
    display: flex;
    flex-direction: column;
}

.name-container {
    display: flex;
    flex-direction: row;
}

.name {
    @extend .font;

    font-size: 46px;
    color: $white;
    margin-right: 30px;
    width: fit-content;
}

.rating {
    @extend .font;

    display: flex;
    border-radius: 50%;
    width: 66px;
    height: 66px;
    color: $light-green;
    border: 1px solid $white;
    align-items: center;
    justify-content: center;
    font-size: 22px;
}

.summary-container {
    margin-top: 30px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 180px;
}

.short-description {
    @extend .film-short-info;

    margin-top: 10px;
    font-size: 16px;
}

.description {
    @extend .font;

    margin-top: 20px;
    font-size: 16px;
    color: $white;
}
</style>
