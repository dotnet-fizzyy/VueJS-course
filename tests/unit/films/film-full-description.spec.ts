import FilmFullDescription from '@/components/films/FilmFullDescription.vue';
import Vue from 'vue';
import { Wrapper, shallowMount } from '@vue/test-utils';
import { createDefaultVueInstance } from '../../setup';

const localVue = createDefaultVueInstance();

describe('FilmFullDescription.vue Tests', () => {
    const createComponent = (propsData): Wrapper<Vue> =>
        shallowMount(FilmFullDescription, {
            localVue,
            propsData,
        });

    const findFilmRating = (wrapper: Wrapper<Vue>) => wrapper.find('[data-aqa-full-rating]');

    const findAllFilmDescriptionItemComponents = (wrapper: Wrapper<Vue>) =>
        wrapper.findAllComponents({ name: 'film-description-item' });

    it('Should render', () => {
        //Arrange
        const wrapper = createComponent({
            name: 'Film Name',
            releaseYear: 2022,
            runtime: 120,
            rating: 4.2,
            shortDescription: 'Short Description',
            fullDescription: 'Some Long Film Description',
            posterUrl: 'posterUrl',
            posterAlt: 'posterAlt',
        });

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should add precision to rating if it is numeral', () => {
        //Arrange
        const numeralRating = 4;

        const wrapper = createComponent({
            name: 'Film Name',
            releaseYear: 2022,
            runtime: 120,
            rating: numeralRating,
            shortDescription: 'Short Description',
            fullDescription: 'Some Long Film Description',
            posterUrl: 'posterUrl',
            posterAlt: 'posterAlt',
        });

        //Act
        const ratingWrapper = findFilmRating(wrapper);

        //Assert
        expect(ratingWrapper.element).toBeTruthy();
        expect(ratingWrapper.text()).toEqual(`${numeralRating}.0`);
    });

    it('Should round to up film runtime if it is digit', () => {
        //Arrange
        const originalRuntime = 121.4;
        const expectedRuntime = 122;

        const wrapper = createComponent({
            name: 'Film Name',
            releaseYear: 2022,
            runtime: originalRuntime,
            rating: 4.2,
            shortDescription: 'Short Description',
            fullDescription: 'Some Long Film Description',
            posterUrl: 'posterUrl',
            posterAlt: 'posterAlt',
        });

        //Act
        const filmDescriptionWrappers = findAllFilmDescriptionItemComponents(wrapper);
        const runtimeWrapper = filmDescriptionWrappers.wrappers[1];

        //Assert
        expect(runtimeWrapper.element).toBeTruthy();
        expect(runtimeWrapper.attributes('additionallabel')).toEqual('min');
        expect(runtimeWrapper.attributes('keylabel')).toEqual(expectedRuntime.toString());
    });
});
