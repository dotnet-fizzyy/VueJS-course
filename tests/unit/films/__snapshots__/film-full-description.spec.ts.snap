// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmFullDescription.vue Tests Should render 1`] = `
<div
  class="root"
  data-aqa-full-description=""
>
  <img
    alt="posterAlt"
    class="poster"
    data-aqa-full-poster=""
  />
   
  <div
    class="film-description-container"
  >
    <div
      class="name-container"
    >
      <span
        class="name"
        data-aqa-full-desc-name=""
      >
        Film Name
      </span>
       
      <span
        class="rating"
        data-aqa-full-rating=""
      >
        4.2
      </span>
    </div>
     
    <span
      class="short-description"
      data-aqa-full-short-desc=""
    >
      Short Description
    </span>
     
    <div
      class="summary-container"
    >
      <film-description-item-stub
        additionallabel="year"
        keylabel="2022"
      />
       
      <film-description-item-stub
        additionallabel="min"
        keylabel="120"
      />
    </div>
     
    <span
      class="description"
      data-aqa-full-full-desc=""
    >
      Some Long Film Description
    </span>
  </div>
</div>
`;
